internal/state/relationship/capturer.go

```diff
@@ -63,32 +84,89 @@ func (c *CapturerImpl) Capture(obj client.Object) {
 				Name:      svcName,
 			}
 		}
+	case *v1beta1.Gateway:
+		var selectors []labels.Selector
+		for _, listener := range o.Spec.Listeners {
+			if selector := graph.GetAllowedRouteLabelSelector(listener); selector != nil {
+				convertedSelector, err := metav1.LabelSelectorAsSelector(selector)
+				if err == nil {
+					selectors = append(selectors, convertedSelector)
+				}
+			}
+		}
+
+		nsname := client.ObjectKeyFromObject(o)
+		if len(selectors) > 0 {
+			c.listenerLabelSelectors[nsname] = selectors
+			for ns, cfg := range c.referencedNamespaces {
+				for _, selector := range selectors {
+					if selector.Matches(labels.Set(cfg.labelMap)) {
+						cfg.match = true
+						cfg.gateways[nsname] = struct{}{}
+						break
+					}
+				}
+				c.referencedNamespaces[ns] = cfg
+			}
+		} else if _, exists := c.listenerLabelSelectors[nsname]; exists {
+			// label selectors existed previously for this gateway, so clean up any references to them
+			c.removeGatewayLabelSelector(nsname)
+		}
+	case *v1.Namespace:
+		nsLabels := o.GetLabels()
+		gateways := c.matchingGateways(nsLabels)
+		nsCfg := namespaceCfg{
+			labelMap: nsLabels,
+			match:    len(gateways) > 0,
+			gateways: gateways,
+		}
+		c.referencedNamespaces[client.ObjectKeyFromObject(o)] = nsCfg
 	}
 }
 
 // Remove removes the relationship for the given object from the CapturerImpl.
-func (c *CapturerImpl) Remove(resourceType client.Object, nsname types.NamespacedName) {
-	switch resourceType.(type) {
+func (c *CapturerImpl) Remove(resource client.Object, nsname types.NamespacedName) {
+	switch resource.(type) {
 	case *v1beta1.HTTPRoute:
 		c.deleteForRoute(nsname)
 	case *discoveryV1.EndpointSlice:
 		delete(c.endpointSliceOwners, nsname)
+	case *v1beta1.Gateway:
+		c.removeGatewayLabelSelector(nsname)
+	case *v1.Namespace:
+		delete(c.referencedNamespaces, nsname)
 	}
 }
 
 // Exists returns true if the given object has a relationship with another object.
-func (c *CapturerImpl) Exists(resourceType client.Object, nsname types.NamespacedName) bool {
-	switch resourceType.(type) {
+func (c *CapturerImpl) Exists(resource client.Object, nsname types.NamespacedName) bool {
+	switch resource.(type) {
 	case *v1.Service:
 		return c.serviceRefCount[nsname] > 0
 	case *discoveryV1.EndpointSlice:
 		svcOwner, exists := c.endpointSliceOwners[nsname]
 		return exists && c.serviceRefCount[svcOwner] > 0
+	case *v1.Namespace:
+		cfg, exists := c.referencedNamespaces[nsname]
+		return exists && cfg.match
 	}
 
 	return false
 }
 
+// RelationshipEnded checks if a relationship previously existed, but no longer does.
+func (c CapturerImpl) RelationshipEnded(resource client.Object) bool {
```

Comment:
Why is this needed for namespaces?

Reply:
If a namespace has its labels removed, calling `Exists` will return false since there's no longer a relationship, which results in no config changes being generated. However, because that namespace is no longer allowed, we need to regenerate our config to remove any routes that reference it. So this function ensures that we push an update in that case.

Reply:
I see. And this is unique to Namespaces? It just makes me wonder about the other relationships we are tracking here.

Reply:
Right now I think this is just unique to namespaces.

Reply:
maybe we should update Capture and Delete so that they return changed if the relationship was changed? For example, if Capture removed a relationship, it will returned changed. If it created a relationship, it will return  changed. This way we will avoid additional methods.

Reply:
That could work.